### PR TITLE
Add discovery of Yeelight devices

### PR DESCRIPTION
**Pull request by valashko**
This pull request makes it possible to discover multiple Yeelight devices and store additional information such as id, firmware version etc.

### CODE REVIEW

**Review by viktorteren — APPROVED** [review]
I tried these changes, works great

**Comment by florian-mollin** [comment]
Thanks for your PR!
As you can see, this repo is a little bit "inactive" and I don't really spend much time maintaining it anymore, but it's an important PR.
I will review your code and make some comments for the "debate" and progress for merge this PR

File: README.md [hunk 1]
```diff
@@ -18,5 +18,11 @@ device.setPower(true);
 device.setRGB(255, 126, 0);
 // Change device brightness
 device.setBrightness(100);
+
+// Discover and instantiate devices
+ArrayList<YeelightDevice> devices = new ArrayList<>();
```
Comment (florian-mollin) [inline README.md:23]: Maybe use `List` instead `ArrayList`

File: src/main/java/com/mollin/yapi/YeelightDevice.java [hunk 2]
```diff
@@ -36,6 +36,10 @@
      * Device effect duration setting for commands
      */
     private int duration;
+    /**
+     * Additional information about a Yeelight device
+     */
+    private Optional<YeelightDeviceMeta> meta;
```
Comment (florian-mollin) [inline src/main/java/com/mollin/yapi/YeelightDevice.java:42]: I'm not a huge fan for a "meta" attribute (for example if you instanciate a Bulb with only IP address, there is no "meta" information)

File: src/main/java/com/mollin/yapi/YeelightDevice.java [hunk 3]
```diff
@@ -310,4 +324,9 @@ public void setName(String name) throws YeelightResultErrorException, YeelightSo
         YeelightCommand command = new YeelightCommand("set_name", name);
         this.sendCommand(command);
     }
+
+    public YeelightDeviceMeta getMeta() {
```
Comment (florian-mollin) [inline src/main/java/com/mollin/yapi/YeelightDevice.java:328]: If we keep "meta", getter must return an Optional and not a default meta with empty attributes

File: src/main/java/com/mollin/yapi/YeelightDeviceMeta.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package com.mollin.yapi;
+
+public class YeelightDeviceMeta {
```
Comment (florian-mollin) [inline src/main/java/com/mollin/yapi/YeelightDeviceMeta.java:3]: Maybe some JavaDoc ?

File: src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetSocketAddress;
+import java.net.SocketTimeoutException;
+import java.util.HashSet;
+import java.util.Set;
+
+public class YeelightDiscoveryManager {
```
Comment (florian-mollin) [inline src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java:11]: I think it's a singleton ? So use design pattern singleton

File: src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    private final static String NEWLINE = "\r\n";
+
```
Comment (florian-mollin) [inline src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java:13]: Why new line here ?

File: src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    private final static String SEARCH_REQUEST_ADDRESS = "239.255.255.250";
+    private final static int SEARCH_REQUEST_PORT = 1982;
+    private final static int RESPONSE_TIMEOUT = 1000;
+
+    public static Set<YeelightDeviceMeta> search() throws IOException {
```
Comment (florian-mollin) [inline src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java:18]: Some javadoc ?

File: src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java (hunk 5 above)
Comment (florian-mollin) [inline src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java:11]: javadoc ?

Comment (florian-mollin) [reply src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java:11]: This code is yours? If no, cite the author (in javadoc ?)

File: src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        final String request = "M-SEARCH * HTTP/1.1"    + NEWLINE +
+                               "MAN: \"ssdp:discover\"" + NEWLINE +
+                               "ST: wifi_bulb"          + NEWLINE;
+        final byte[] requestBytes = request.getBytes();
+
+        HashSet<YeelightDeviceMeta> devices = new HashSet<>(); // prevent duplicates as a result of multiple responses from the same device
```
Comment (florian-mollin) [inline src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java:24]: Use `Set` instead `HashSet`

File: src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+        InetSocketAddress address = new InetSocketAddress(SEARCH_REQUEST_ADDRESS, SEARCH_REQUEST_PORT);
+        DatagramPacket requestPacket = new DatagramPacket(requestBytes, requestBytes.length, address);
+        DatagramSocket socket = new DatagramSocket();
+        socket.setSoTimeout(RESPONSE_TIMEOUT);
+        socket.send(requestPacket);
+        try {
+            while(true) {
+                byte[] responseBuffer = new byte[1500];
```
Comment (florian-mollin) [inline src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java:33]: `1500` ?

File: src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+                devices.add(device);
+            }
+        } catch (SocketTimeoutException e) {
+            // signifies that all responses have been processed
+        }
+        return devices;
+    }
+
+    private static YeelightDeviceMeta processResponse(String response) {
+        // TODO response validation
```
Comment (florian-mollin) [inline src/main/java/com/mollin/yapi/YeelightDiscoveryManager.java:46]: TODO here

**Review by florian-mollin — CHANGES_REQUESTED** [review]
- There are some changes to be made (see review).
- Maybe remove `meta` attribute from `YeelightDevice` class (user can retrieve information with your `YeelightDeviceMeta` class)
- Maybe add some tests as far as possible
- I have not yet tested the code

**Comment by florian-mollin** [comment]
It's been a long time, I think I can close this PR

**Pull request closed without merging**